### PR TITLE
added limited support for analytic kl divergences for elbos

### LINKED ISSUE #688: Use kl_divergence() in SVI
PyTorch distributions implements a `kl_divergence()` function. We should use this when available in SVI. First we need to account for Pyro distribution wrappers:
```py
def kl_divergence(p, q):
    try:
        p = p.torch_dist
        q = q.torch_dist
    except AttributeError:
        raise NotImplementedError
    return torch.distributions.kl_divergence(p, q)
```
then in SVI we can `try...except`:
```py
assert self.assume_mean_field and self.use_analytic_kl
try:
    kl = kl_divergence(guide_site['fn'], model_site['fn'])
    elbo += kl.detach().sum()
    surrogate_elbo += kl
except NotImplementedError:
    # fall back to monte carlo algorithm
```

### PR DESCRIPTION
**Pull request by martinjankowiak**
Resolves #688
pair coded with @fritzo.

## summary

this adds support for an `SVI` object that supports analytic kl divergences. this `ELBO`
assumes that:

- everything is reparameterized (have checks for this)
- a certain dependency condition is satisfied. we have an approximate check for this, and describe the exact requirement in the docstring. since mean field guides is a common case in which this condition is satisfied, we use mean field as the name of the class

## reviewers note:
- ~~do we need any more tests?~~
- ~~i removed some `xfail_param` statements in test_gradient. if it is important to reinstate these please let me know how to do so (i couldn't figure it out).~~

## tested

- [x] gradient tests
- [x] autoguide tests with auxiliary guide variables
- [x] death tests for warning about invalid model structure

### CODE REVIEW

**Comment by martinjankowiak** [comment]
@fritzo do you want to add the masked support directly into this PR?

**Comment by fritzo** [comment]
@martinjankowiak one of the autoguide tests is failing with `TraceMeanField_ELBO`, can you help diagnose?

File: pyro/infer/trace_mean_field_elbo.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    Note for advanced users:
+
+    The mean field condition is a sufficient but not necessary condition for
+    this estimator to be correct. The precise condition is that for every
+    latent variable `z` in the guide, its parents in the model must not include
+    any latent variables that are descendants of `z` in the guide. Here
+    'parents in the model' and 'descendants in the guide' is with respect
+    to the corresponding (statistical) dependency structure. For example, this
+    condition is always satisfied if the model and guide have identical
+    dependency structures.
```
Comment (fritzo) [inline pyro/infer/trace_mean_field_elbo.py]: Would it be sufficient and not too inconvenient to require the model and guide have the same variable ordering, and warn otherwise? This is easy to check from the traces.

Comment (fehiepsi) [reply pyro/infer/trace_mean_field_elbo.py]: @fritzo I guess the trace just gives the "coding" order, not "structure" order.

Comment (martinjankowiak) [reply pyro/infer/trace_mean_field_elbo.py]: @fritzo i think that would be reasonable. probably warn over error/assert

**Comment by martinjankowiak** [comment]
@fritzo not sure what's going on with the autoguide test (we should take a look together), but looking at this makes me realize that this implementation will break for subsampled latent variables (no problems with subsampled observations)

File: pyro/infer/trace_mean_field_elbo.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+def _check_mean_field_requirement(model_trace, guide_trace):
+    """
+    Checks that the guide and model sample sites are ordered identically.
+    This is sufficient but not necessary for correctness.
+    """
+    model_sites = [name for name, site in model_trace.nodes.items()
+                   if site["type"] == "sample" and name in guide_trace.nodes]
+    guide_sites = [name for name, site in guide_trace.nodes.items()
+                   if site["type"] == "sample" and name in model_trace.nodes]
+    assert set(model_sites) == set(guide_sites), "programmer error"
```
Comment (martinjankowiak) [inline pyro/infer/trace_mean_field_elbo.py]: 'programmer error' seems a potentially confusing assert msg, no?

Comment (fritzo) [reply pyro/infer/trace_mean_field_elbo.py]: if find it standard, and it should never be surfaced to users, but ok I'll remove it.

**Comment by martinjankowiak** [comment]
might need to adjust the precision on failing test down a bit?

**Comment by fehiepsi** [comment]
Love the latest change, we only check things when validation is enable.

**Comment by fritzo** [comment]
@gpleiss hopefully this will speed up your GPyTorch+Pyro models. Let us know if you have any issues!
```diff
- SVI(model, guide, optim, Trace_ELBO())
+ SVI(model, guide, optim, TraceMeanField_ELBO())
```

**Comment by gpleiss** [comment]
Awesome! Excited to try it out :)

**Comment by royf** [comment]
Very cool, thanks!

Are there plans to support non-reparametrizable variables?
I have a use case where variables cannot be reparametrized, and where training without analytic KL cannot work.
So I cannot use Pyro until it supports analytic KL of non-reparametrizable variables.
